internal/impl/kafka/franz_shared_client.go

```diff
@@ -0,0 +1,127 @@
+)
+
+var (
+	errSharedClientNameDuplicate = errors.New("a duplicate name for shared clients has been detected")
+	errSharedClientNameNotFound  = errors.New("shared client not found")
+)
+
+// FranzSharedClientSet attempts to store a shared client with a given
+// identifier in the provided resources pointer.
+func FranzSharedClientSet(name string, client *FranzSharedClientInfo, res *service.Resources) error {
+	reg := getSharedClientRegister(res)
+	return reg.set(name, client)
+}
+
+// FranzSharedClientPop attempts to remove and return a shared client with a
+// given identifier in the provided resources pointer.
+func FranzSharedClientPop(name string, res *service.Resources) (*FranzSharedClientInfo, error) {
+	reg := getSharedClientRegister(res)
+	return reg.pop(name)
+}
+
+// FranzSharedClientUseFn defines a closure that receives shared client details.
+type FranzSharedClientUseFn func(details *FranzSharedClientInfo) error
+
+// FranzSharedClientUse attempts to access a shared client with a given
+// identifier in the provided resources pointer.
+func FranzSharedClientUse(name string, res *service.Resources, fn FranzSharedClientUseFn) error {
+	reg := getSharedClientRegister(res)
+	return reg.use(name, fn)
+}
+
+// FranzSharedClientInfo provides an active client and the connection details
+// used to create it.
+type FranzSharedClientInfo struct {
+	Client      *kgo.Client
+	ConnDetails *FranzConnectionDetails
+}
+
+//------------------------------------------------------------------------------
+
+type franzSharedClientRegister struct {
+	mut     sync.RWMutex
+	clients map[string]*FranzSharedClientInfo
+}
+
+func (r *franzSharedClientRegister) set(name string, client *FranzSharedClientInfo) error {
+	r.mut.Lock()
+	defer r.mut.Unlock()
+
+	if r.clients == nil {
+		r.clients = map[string]*FranzSharedClientInfo{}
+	}
+
+	_, exists := r.clients[name]
+	if exists {
+		return errSharedClientNameDuplicate
+	}
+
+	r.clients[name] = client
+	return nil
+}
+
+func (r *franzSharedClientRegister) pop(name string) (*FranzSharedClientInfo, error) {
+	r.mut.Lock()
+	defer r.mut.Unlock()
+
+	if r.clients == nil {
+		return nil, errSharedClientNameNotFound
+	}
+
+	e, exists := r.clients[name]
+	if !exists {
+		return nil, errSharedClientNameNotFound
+	}
+
+	delete(r.clients, name)
+	return e, nil
+}
+
+func (r *franzSharedClientRegister) use(name string, fn func(*FranzSharedClientInfo) error) error {
+	r.mut.RLock()
+	defer r.mut.RUnlock()
+
+	if r.clients == nil {
+		return errSharedClientNameNotFound
+	}
+
+	e, exists := r.clients[name]
+	if !exists {
+		return errSharedClientNameNotFound
+	}
+
+	return fn(e)
+}
+
+//------------------------------------------------------------------------------
+
+type franzSharedClientKeyType int
+
+var franzSharedClientKey franzSharedClientKeyType
+
+func getSharedClientRegister(res *service.Resources) *franzSharedClientRegister {
+	reg, _ := res.GetOrSetGeneric(franzSharedClientKey, &franzSharedClientRegister{})
```

Comment:
It's to avoid a map allocation every time we get a shared client handle. I'm not actually sure what the real life impact would be, it's probably negligible but I'd rather keep it as is for now, will add a comment though.

Reply:
Oh, I didn't think of that. Good point! ❤️